### PR TITLE
WARMFIX [DEV-4963] Sanitize elasticsearch text input

### PR DESCRIPTION
**Pull request by emily-brents**
**Description:**
Adds sanitization for elasticsearch text input to advanced search filters.

**Technical details:**
Strips all lucene special characters from the text entered into the advanced search filters. This is slightly different from the keyword search method, which strips only specific characters as the keyword search functionality allows using certain characters.

Additionally, for whatever reason importing the sanitization function from elasticsearch_helpers.py would not work in query_with_filters.py so we had to redefine it there.

**Requirements for PR merge:**

1. [X] Unit & integration tests updated
2. [X] API documentation updated
3. [x] Necessary PR reviewers:
    - [x] Backend
4. [N/A] Matview impact assessment completed
5. [N/A] Frontend impact assessment completed
6. [X] Data validation completed
7. [N/A] Appropriate Operations ticket(s) created
8. [X] Jira Ticket [DEV-4963](https://federal-spending-transparency.atlassian.net/browse/DEV-4963):
    - [ ] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
No changes to matviews, no ops tickets needed
```

### CODE REVIEW

**Comment by emily-brents** [comment]
> I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.

Most of the advanced search filters do not allow text input, so it's redundant to sanitize them. Additionally, not all the filters use text values, so we can't just sanitize them all the same way without it causing errors

**Comment by emily-brents** [comment]
> I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.

**Comment by emily-brents** [comment]
> I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.

So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.

**Comment by tony-sappe** [comment]
> > I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.
> 
> So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.

Put another way: the intention for this ticket is to sanitize user-input text fields on the website's Advanced Search page.

File: usaspending_api/search/v2/es_sanitization.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import logging
```
Comment (tony-sappe) [inline usaspending_api/search/v2/es_sanitization.py:1]: If `es_sanitize` is being segregated, would it be agreeable to move both it and `es_minimal_sanitize` together in the `usaspending-api/usaspending_api/common/elasticsearch/` directory?

**Comment by BrianZito** [comment]
> > > I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.
> > 
> > 
> > So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.
> 
> Put another way: the intention for this ticket is to sanitize user-input text fields on the website's Advanced Search page.

Aren't all of these fields user-input by an API user? As long as non-text fields don't need to be sanitized this still shouldn't be a problem though.

File: usaspending_api/search/v2/es_sanitization.py (hunk 1 above)
Comment (emily-brents) [reply usaspending_api/search/v2/es_sanitization.py:1]: that's very agreeable

**Comment by tony-sappe** [comment]
> > > > I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.
> > > 
> > > 
> > > So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.
> > 
> > 
> > Put another way: the intention for this ticket is to sanitize user-input text fields on the website's Advanced Search page.
> 
> Aren't all of these fields user-input by an API user? As long as non-text fields don't need to be sanitized this still shouldn't be a problem though.

Technically yes, all values provided to the API are "user-input", which is why the previous comment was clear on the website user-provided text fields. The ticket was written to address specific errors noticed in the production log statements. It is a warm-fix to ensure it reaches production ahead of "flipping the switch" to Elasticsearch.

**Comment by BrianZito** [comment]
> > > > > I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.
> > > > 
> > > > 
> > > > So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.
> > > 
> > > 
> > > Put another way: the intention for this ticket is to sanitize user-input text fields on the website's Advanced Search page.
> > 
> > 
> > Aren't all of these fields user-input by an API user? As long as non-text fields don't need to be sanitized this still shouldn't be a problem though.
> 
> Technically yes, all values provided to the API are "user-input", which is why the previous comment was clear on the website user-provided text fields. The ticket was written to address specific errors noticed in the production log statements. It is a warm-fix to ensure it reaches production ahead of "flipping the switch" to Elasticsearch.

Is there a ticket for the API user-provided text fields that's going to be a warmfix? If we're looking at a vulnerability we need to patch it at both points.

**Comment by tony-sappe** [comment]
> > > > > > I still see several filters that don't have sanitization. As for those that follow the standard list of strings format for filter_values, I would rather see the sanitization done in advance, so that the onus isn't on the implementer of each filter to remember this requirement. I wrote the `pre-sanitize` branch to demonstrate this. Didn't test it though.
> > > > > 
> > > > > 
> > > > > So sanitizing all the fields the same won't work, since not all the fields are plain text fields some are dicts, some are. Even if we did want to go through all the dictionaries to sanitize them, we wouldn't actually want to do that, since cleaning the text input requires removing characters that work in elasticsearch for non-text fields, so it would break the time period filter entirely. I can go through an add the sanitization to fields that won't break, but it would still be mostly redundant for any field that isn't direct text input, which is only Keyword, recipient, and award id.
> > > > 
> > > > 
> > > > Put another way: the intention for this ticket is to sanitize user-input text fields on the website's Advanced Search page.
> > > 
> > > 
> > > Aren't all of these fields user-input by an API user? As long as non-text fields don't need to be sanitized this still shouldn't be a problem though.
> > 
> > 
> > Technically yes, all values provided to the API are "user-input", which is why the previous comment was clear on the website user-provided text fields. The ticket was written to address specific errors noticed in the production log statements. It is a warm-fix to ensure it reaches production ahead of "flipping the switch" to Elasticsearch.
> 
> Is there a ticket for the API user-provided text fields that's going to be a warmfix? If we're looking at a vulnerability we need to patch it at both points.

No. This PR is the warmfix under [DEV-4963](https://federal-spending-transparency.atlassian.net/browse/DEV-4963). For several weeks 100% of Advanced Search visualization API requests were forked and sent to Elasticsearch. A high number of Elasticsearch query errors were directly due to the keyword filter text values.

Sanitizing all inputs is a good idea, that's where TinyShield comes into play. But this PR and the ticket is specifically to address the errors occurring now.

**Pull request merged**